📖 Condition CAEP

fabriziopandini opened the pull request:
**What this PR does / why we need it**:
This PR brings the CAEP for introducing conditions in Cluster API

The discussion on the Google doc are addressed, with expectation for:
- Confirming/Removing the `Severity` field
- Adding the `ObservedGerneration` field

Moving here for getting feedback from a broader adience

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
xref https://github.com/kubernetes-sigs/cluster-api/issues/3005

/assing @vincepri 
/assing @ncdc

vincepri commented:
/hold

for community consensus

ncdc commented on docs/proposals/20200506-conditions.md:
```diff
@@ -0,0 +1,363 @@
+
+Condition types MUST have one of the following suffix:
+
+- `Ready`, for resources which represent an ongoing status, like `ControlplaneReady` or `MachineDeploymentsReady`.
```
```suggestion
- `Ready`, for resources which represent an ongoing status, like `ControlPlaneReady` or `MachineDeploymentsReady`.
```

ncdc commented on docs/proposals/20200506-conditions.md:
```diff
@@ -0,0 +1,363 @@
+worst case, you have to look at the logs from the Cluster API pods to get more details on why
+things aren’t working.
+
+This proposal introduces the concept of “conditions” as a tool designed to provide “at glance”
```
```suggestion
This proposal introduces the concept of “conditions” as a tool designed to provide an “at a glance”
```

vincepri commented on docs/proposals/20200506-conditions.md:
```diff
@@ -0,0 +1,363 @@
+4. To provide flexibility for providers to incrementally adopt this proposal; or add, remove,
+   or update their own set of conditions.
+
+Once there will be an agreed design for the condition types, a second iteration will start
+with the following goals:
+
+1. To define the first set of conditions on the different Cluster API objects.
+2. To have `clusterctl describe cluster` present an overview of the cluster, its infrastructure,
+  control plane, and other related objects.
+
+The final iteration of this proposal will focus on:
+
+1. To define a mechanism for bubbling up/aggregating conditions up to hierarchy of cluster
+   objects
```
Should we move these points to the non-goals/future work section?

fabriziopandini replied:
done

vincepri commented on docs/proposals/20200506-conditions.md:
```diff
@@ -0,0 +1,363 @@
+- To remove  `Phase`, `ErrorMessage`, `ErrorReason` from Status in v0.3 .
+- To modify any Cluster API code in v0.3 to rely on conditions to determine actions or workflows.
+
+Future Work (v1alpha4)
```
I'd just have one unified list, like in other proposals.

fabriziopandini replied:
done

vincepri commented on docs/proposals/20200506-conditions.md:
```diff
@@ -0,0 +1,363 @@
+Non-Goals
+
+- To pre-define every possible condition.
+- To remove  `Phase`, `ErrorMessage`, `ErrorReason` from Status in v0.3 .
```
```suggestion
- To remove `Phase`, `ErrorMessage`, `ErrorReason` from Status in v0.3 .
```

vincepri commented on docs/proposals/20200506-conditions.md:
```diff
@@ -0,0 +1,363 @@
+Once there will be an agreed design for the condition types, a second iteration will start
+with the following goals:
+
+1. To define the first set of conditions on the different Cluster API objects.
+2. To have `clusterctl describe cluster` present an overview of the cluster, its infrastructure,
+  control plane, and other related objects.
```
Shouldn't these to be part of the first iteration?

fabriziopandini replied:
Considered the fact that we are anticipating the implementation of the process for making conditions visible thought the hierarchy, I'm putting this goal on hold

vincepri commented on docs/proposals/20200506-conditions.md:
```diff
@@ -0,0 +1,363 @@
+
+## Proposal
+
+Disclaimer: the goal of this first iteration is to to agree on a design for the condition types. 
```
```suggestion
```

This is clear from the goal section

vincepri commented on docs/proposals/20200506-conditions.md:
```diff
@@ -0,0 +1,363 @@
+
+### User Stories
+
+- Detail the things that people will be able to do if this proposal is implemented.
+- Include as much detail as possible so that people can understand the "how" of the system.
+- The goal here is to make this feel real for users without getting bogged down.
```
Should we remove this boilerplate?

fabriziopandini replied:
yes!

vincepri commented on docs/proposals/20200506-conditions.md:
```diff
@@ -0,0 +1,363 @@
+- Include as much detail as possible so that people can understand the "how" of the system.
+- The goal here is to make this feel real for users without getting bogged down.
+
+#### Story 1
```
I was expecting to see more user stories, including end user perspectives

fabriziopandini replied:
done!

vincepri commented on docs/proposals/20200506-conditions.md:
```diff
@@ -0,0 +1,363 @@
+
+#### Data Model Changes
+
+In this section we outline the proposed API changes.
```
```suggestion
In this section we outline the proposed API additions.
```

vincepri commented on docs/proposals/20200506-conditions.md:
```diff
@@ -0,0 +1,363 @@
+   Status corev1.ConditionStatus `json:"status"`
+
+   // Severity with which to treat failures of this type of condition.
+   // When this is not specified, it defaults to Error.
```
Why not make it a required field? Defaulting to `Error` might result in unexpected consequences

fabriziopandini replied:
Severity is relevant only for failed conditions

vincepri commented on docs/proposals/20200506-conditions.md:
````diff
@@ -0,0 +1,363 @@
+error or to other events that are transient or can be eventually remediated by an user intervention
+
+```
+MachineReady=False, Reason=MachineNotHeathly, Severity=Error
````
```suggestion
MachineReady=False, Reason=MachineNotHealthy, Severity=Error
```

vincepri commented on docs/proposals/20200506-conditions.md:
```diff
@@ -0,0 +1,363 @@
+
+### Risks and Mitigations
+
+- Risk: This proposal is expected to be completed in three iterations, but in the initial stages this can make it
+  harder for the reviewers to figure out what will be the final result of this effort.
+- Mitigation: The incremental approach already embeds the possibility to make adjustments along the way. 
+
+- Risk: This proposal aims to be consistent with the target state of conditions in Kubernetes, but this is still
+  under definition (see [KEP](https://github.com/kubernetes/enhancements/pull/1624)).
+- Mitigation: Periodically re-evaluate this proposal vs the Kubernetes KEP. 
+
+- Risk: Cluster API presents some specific challenges that are not common to the core Kubernetes objects.
+- Mitigation: To allow a minimal set of carefully evaluated differences between Cluster API and Kubernetes
+  conditions.
```
A lot of these points are repeated above, should we dedupe?

fabriziopandini replied:
Done

vincepri commented on docs/proposals/20200506-conditions.md:
```diff
@@ -0,0 +1,363 @@
+
+Condition types MUST have a consistent polarity (i.e. "True = good");
+
+Condition types MUST have one of the following suffix:
```
How would we express things like `Timeouts`, for example from MachineHealthCheck

fabriziopandini replied:
IMO a missed timeout is a reason for a condition failing, not a separate condition

enxebre commented on docs/proposals/20200506-conditions.md:
```diff
@@ -0,0 +1,363 @@
+Once there will be an agreed design for the condition types, a second iteration will start
+with the following goals:
+
+1. To define the first set of conditions on the different Cluster API objects.
```
I think this needs to be conflated with the first interaction. In order to come up with a sane data model we need to have first a clear understanding and view for each component needs.

fabriziopandini replied:
Agreed. Done
however, even if we are conflating iterations, please note those objectives are still non-goals
- To pre-define every possible condition.
- To define the exact semantic for every condition introduced in the proposal 
This should be done by implementation PR once there is an agreement on data structure and on constraints/design principles for defining conditions.

enxebre commented on docs/proposals/20200506-conditions.md:
```diff
@@ -0,0 +1,363 @@
+All the changes can be implemented in the v0.3.x timeframe (no breaking changes, only additions).
+
+Following types should be defined in Cluster API / v1alpha3 api.
+
```
I think that to figure out a data model and possibly justify skew from https://github.com/kubernetes/enhancements/pull/1624 we need to first flesh out some user stories traversing the different CAPI components to understand which data model can best convey our contents.

fabriziopandini replied:
Done, user stories and examples are now included

enxebre commented on docs/proposals/20200506-conditions.md:
```diff
@@ -0,0 +1,363 @@
+
+The goal of this first iteration is:
+
+1. To agree on a design for the condition types.
```
I think that to agree on a design for the condition types we need to include as goals a list of user stories traversing capi components which are not well communicated today but they will with the new conditions. Then in the "Implementation" section we can elaborate how we'll achieve these goals fleshing out where we set those conditions, which values and how the user can see them in a meaningful manner.

fabriziopandini replied:
I think we should preserve the distinction between goals and user stories, but I agree on expanding the implementation section with user stories and examples traversing CAPI components (see new commits)

davidewatson commented on docs/proposals/20200506-conditions.md:
```diff
@@ -0,0 +1,363 @@
+The `Severity` field MUST be set only when `Status=False` and it is designed to provide a standard classification
+of possible conditions failure `Reason`. 
+
+Please note that the combination of `Reason` and `Severity` gives different meaning to a condition failure
```
Given my comment above about what a condition is, I wonder if `Severity` might be the wrong word, or even mechanism, for classifying conditions. The information encoded there (i.e. `Info`, `Warning`, `Error`) comes from logging, and even there [some people dispute](https://dave.cheney.net/2015/11/05/lets-talk-about-logging) whether there is a need for more than `Info` and `Error`. Is there a more specific word, which applies to all conditions, which we can use? Or should we just treat Conditions as freeform messages and force callers to know which conditions they care about ahead of time? Could we have a convention for encoding whether a Condition is terminal or not in `Type`?

I would also be curious if this discussion has come up in SIG API Machinery...

Sorry if this note is confusing at all! I probably need to review @bgrant0607's [conditions references](https://github.com/kubernetes/enhancements/pull/1624#issuecomment-603313342) in more detail...

fabriziopandini replied:
> The information encoded there (i.e. Info, Warning, Error) comes from logging

True, but this is also a generally understood way to express:
- You are good
- You should take a look
- You have some problems

> I wonder if Severity might be the wrong word

I have considered `ReasonClass`, `ReasonCategory`, but at the end opted for `Severity` because the focus here is in catching the user's attention.
But I'm more than happy to hear suggestions here...

> should we just treat Conditions as freeform messages and force callers to know which conditions they care about ahead of time

free form message is difficult for consumers, so we are encoding ConditionTypes, Reason, and Severity to make it easier to deal with conditions

fabriziopandini commented on docs/proposals/20200506-conditions.md:
```diff
@@ -0,0 +1,363 @@
+  ConditionSeverityNone ConditionSeverity = ""
+)
+
+// Conditions define a readiness condition for a Cluster API resource.
```
100% agreed

fabriziopandini commented:
@davidewatson @enxebre @ncdc @vincepri thanks for the feedback!
I have updated the proposal adding use cases and examples as requested; as a consequence,
 we are now aiming to a single iteration of the proposal.

vincepri commented:
Holding for 1 more week for async consensus

benmoss commented:
If we are disagreeing with the advice of https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1623-standardize-conditions we should have something saying why

benmoss commented:
The upstream conditions KEP has merged: https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1623-standardize-conditions

ncdc commented on docs/proposals/20200506-conditions.md:
```diff
@@ -0,0 +1,550 @@
+   Severity ConditionSeverity `json:"severity,omitempty"`
+
+   // LastTransitionTime is the last time the condition transitioned from one status to another.
+   // +optional
```
Should we consider making this required instead?

fabriziopandini replied:
Agreed, also on Kubernetes KEP the `LastTransitionTime ` switched to required before merging the KEP

benmoss commented on docs/proposals/20200506-conditions.md:
````diff
@@ -0,0 +1,550 @@
+)
+```
+
+Condition types MUST have a consistent polarity (i.e. "True = good");
````
Why does this matter?

fabriziopandini commented on docs/proposals/20200506-conditions.md:
```diff
@@ -0,0 +1,550 @@
+
+- In K8s the `Type` field is a string, while in this proposal is a string alias so we can enforce the
+  definition of a well-defined set of `ConditionType` constants.
+- In K8s there is an `ObservedGeneration` field, while in this proposal not.
```
My major concern here is that `ObservedGeneration` creates a dependency between `Status.Conditions` conditions to the `ObjectMeta` struct and this makes harder to implement a generic library for handling conditions on all objects.
I know this is an implementation detail, but at this stage, I prefer to give priority to provide a shared library for conditions vs introducing the `ObservedGeneration` field...

vincepri replied:
We can always add it later if we find a clean way with reflection

benmoss replied:
As the caller of the method I probably have access to the `ObjectMeta`, maybe the problem is just that you're trying to put the shared library on the Conditions object

fabriziopandini commented on docs/proposals/20200506-conditions.md:
```diff
@@ -0,0 +1,550 @@
+### User Stories
+
+#### Story 1
+As a developer, as a user, as a tool built on top of Cluster API, I would like to have common types
+for defining conditions on different Cluster API or provider objects.
```
Because all of them could benefit from if the same data type is used across all the Cluster API objects. 
BTW this is the same motivation for https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1623-standardize-conditions, so there is alignment with Kubernetes here

benmoss replied:
That makes sense, maybe it's more a motivation than a user story, but it's fine.

benmoss commented on docs/proposals/20200506-conditions.md:
```diff
@@ -0,0 +1,550 @@
+   LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`
+
+   // The reason for the condition's last transition.
+   // Reasons should be one-word CamelCase.
```
I think this just should say "Reasons should be CamelCase". As written it sounds like they are supposed to be one word, which isn't true given the examples (`MachineNotHealthy` is three words). I think the fact that it's camelcase implies it's joined together without spaces.

detiber commented on docs/proposals/20200506-conditions.md:
````diff
@@ -0,0 +1,550 @@
+)
+```
+
+Condition types MUST have a consistent polarity (i.e. "True = good");
````
While I agree with this in concept, I'm not sure we would want to force this in practice.

In particular, I'm wondering about how this representation would affect the way that we deal with the absence of a given condition, particularly in the case of MachineHealthCheck compared to say something like ClusterInfrastructureReady.

In the case of a missing condition related to the Cluster Infrastructure being ready we would want to consider the default in the case of the condition being not present to be "not ready"(false/unknown) and block any operations that would want Cluster Infrastructure Ready.

However, in the case of MachineHealthCheck if we had a condition that represented an MHC passing (good), we would need to treat the absence of the condition as being true to avoid attempting to perform remediation prior to the MHC running. Not only does this seem a bit counter-intuitive on the remediation side (assuming success in the event of a missing condition) it also seems counter-intuitive from an implementation side for MHC as well, since we cannot really be confident that it is healthy, just that the conditions we are checking for have not failed.

benmoss replied:
Yes, we already are facing this problem in #3056 with the verbal contortions needed to make a condition that means "remediation is needed". We couldn't come up with anything better than `OwnerRemediationSucceeded` being set to false upon failure, which is creating some confusing semantics.

fabriziopandini replied:
@detiber @benmoss 
It is a trade-off.
I'm +1 for trying to enforce this rule because having a consistent polarity ensures a simple and clear contract with the consumers and it allows to manipulate conditions in a simple and consistent way without being forced to implements specific logic for each condition type. 

> I'm wondering about how this representation would affect the way that we deal with the absence of a given condition

IMO a missing condition is unknown (not True, not False); also in this case, a consistent polarity can help in reading code.

fabriziopandini commented:
@vincepri @ncdc @benmoss @detiber I have addressed comment about risks of using a consistent polarity and relaxed naming convention rules, PTAL.

WRT to the management "Unknown" state, I have found a consideration in the recent Kubernetes KEP that IMO clearly defines the problem:

> "Unknown" is a fact about the writer of the condition and not a claim about the object

I agree also on the proposed remediation, which is to require the condition author to clearly document the meaning of the "Unknown" state for each condition.

TL;DR. event if "Unknown" might vary case by case and we should document this, consistent polarity clarifies the meaning of "True" and "False" and this is an improvement vs having different interpretations for all the three possible condition states.

vincepri commented:
Before merging, let's squash and change the status of the proposal to `implementable`

vincepri commented:
/milestone v0.3.x

fabriziopandini commented:
> let's squash and change the status of the proposal to implementable

done

detiber commented on docs/proposals/20200506-conditions.md:
```diff
@@ -0,0 +1,573 @@
+  - `KubeadmControlPlane.Status.Conditions[MachinesReady]` is a condition based on the aggregation
+  of the underlying `Machines.Status.Conditions[Ready]` conditions.
+- A corollary of the above set of constraints is that an object SHOULD NEVER be in status `Ready=True`
+  if one of the object's conditions are `false` or if one of the object dependents is in status `Ready=False`.
```
Does this mean that `Cluster.Status.Conditions[Ready]` will be False if any Machine belonging to the Cluster has `Machine.Status.Conditions[Ready]` = False? 

If so, I'm not sure that I fully agree there, since the Cluster can be fully functional even if some subset of the components that make up the Cluster are not (and for any sufficiently large cluster a degraded state of some kind would be more normal than it would be exceptional).

fabriziopandini replied:
@detiber, I agree with your concern about the need of finding a good rule for what we report all the machines at cluster level, and this is why later in the document, when we discuss WorkerReady, we are stating that "_Considering the complexity and to wait for more in-depth analysis and use-cases, the implementation of this Condition is deferred until a later stage_."

With that in mind, we are embedding following flexibility point in the document:
- each object SHOULD (instead if MUST) have a ready condition (so e.g. during implementation we can choose to not create an overall Ready condition at cluster level)
- the Ready condition state SHOULD (instead if MUST) be consistent with the state of underlying source conditions (so in future there is room to qualify "a degraded state of some kind", like e.g. Info if failures < x%, Warning if ... etc.)

detiber commented:
/lgtm
